### ISSUE TITLE
Sequelize integration failure due to postgres integration

### ISSUE THREAD

**Issue by vados-scout** [scoutapp/scout_apm_node#191]
Current NodeJS implementation is breaking `sequelize`:

```js
{
  "errors": [
    {
      "message": "Cannot read property 'rows' of undefined",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "currentMember"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "TypeError: Cannot read property 'rows' of undefined",
            "    at /users/user/path/to/project/node_modules/sequelize/lib/dialects/postgres/connection-manager.js:286:32",
            "    at propagateAslWrapper (/users/user/path/to/project/node_modules/async-listener/index.js:504:23)",
            "    at /users/user/path/to/project/node_modules/async-listener/glue.js:188:31",
            "    at /users/user/path/to/project/node_modules/async-listener/index.js:541:70",
            "    at /users/user/path/to/project/node_modules/async-listener/glue.js:188:31",
            "    at processTicksAndRejections (internal/process/task_queues.js:97:5)",
            "From previous event:",
            "    at ConnectionManager.connect (/users/user/path/to/project/node_modules/sequelize/lib/dialects/postgres/connection-manager.js:195:8)",
            "    at /users/user/path/to/project/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:318:50",
            "From previous event:",
            "    at ConnectionManager._connect (/users/user/path/to/project/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:318:8)",
            "    at Object.create (/users/user/path/to/project/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:136:28)",
            "    at Pool._createResource (/users/user/path/to/project/node_modules/sequelize-pool/lib/Pool.js:255:8)",
            "    at Pool._dispense (/users/user/path/to/project/node_modules/sequelize-pool/lib/Pool.js:240:12)",
            "    at Pool.acquire (/users/user/path/to/project/node_modules/sequelize-pool/lib/Pool.js:330:10)",
            "    at /users/user/path/to/project/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:287:24",
            "    at Immediate.<anonymous> (/users/user/path/to/project/node_modules/async-listener/glue.js:188:31)",
            "    at processImmediate (internal/timers.js:456:21)",
            "    at process.topLevelDomainCallback (domain.js:137:15)",
            "From previous event:",
            "    at ConnectionManager.getConnection (/users/user/path/to/project/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:286:20)",
            "    at /users/user/path/to/project/node_modules/sequelize/lib/sequelize.js:640:36",
            "From previous event:",
            "    at /users/user/path/to/project/node_modules/sequelize/lib/sequelize.js:631:53",
            "    at /users/user/path/to/project/node_modules/retry-as-promised/index.js:70:21",
            "    at new WrappedPromise (/users/user/path/to/project/node_modules/async-listener/es6-wrapped-promise.js:13:18)",
            "    at retryAsPromised (/users/user/path/to/project/node_modules/retry-as-promised/index.js:60:10)",
            "    at /users/user/path/to/project/node_modules/sequelize/lib/sequelize.js:631:30",
            "From previous event:",
            "    at Sequelize.query (/users/user/path/to/project/node_modules/sequelize/lib/sequelize.js:580:23)",
            "    at QueryInterface.select (/users/user/path/to/project/node_modules/sequelize/lib/query-interface.js:1121:27)",
            "    at /users/user/path/to/project/node_modules/sequelize/lib/model.js:1757:34",
            "    at Immediate.<anonymous> (/users/user/path/to/project/node_modules/async-listener/glue.js:188:31)",
            "    at processImmediate (internal/timers.js:456:21)",
            "    at process.topLevelDomainCallback (domain.js:137:15)",
            "From previous event:",
            "    at Function.findAll (/users/user/path/to/project/node_modules/sequelize/lib/model.js:1755:8)",
            "    at Function.findOne (/users/user/path/to/project/node_modules/sequelize/lib/model.js:1924:17)",
            "    at /users/user/path/to/project/src/resolvers/query/current_member.ts:4:41",
            "    at Generator.next (<anonymous>)",
            "    at /users/user/path/to/project/src/resolvers/query/current_member.ts:8:71",
            "    at new WrappedPromise (/users/user/path/to/project/node_modules/async-listener/es6-wrapped-promise.js:13:18)",
            "    at __awaiter (/users/user/path/to/project/src/resolvers/query/current_member.ts:4:12)",
            "    at resolverFn (/users/user/path/to/project/src/resolvers/query/current_member.ts:3:56)",
            "    at field.resolve (/users/user/path/to/project/node_modules/graphql-extensions/src/index.ts:274:18)",
            "    at resolveFieldValueOrError (/users/user/path/to/project/node_modules/graphql/execution/execute.js:467:18)",
            "    at resolveField (/users/user/path/to/project/node_modules/graphql/execution/execute.js:434:16)",
            "    at executeFields (/users/user/path/to/project/node_modules/graphql/execution/execute.js:275:18)",
            "    at executeOperation (/users/user/path/to/project/node_modules/graphql/execution/execute.js:219:122)",
            "    at executeImpl (/users/user/path/to/project/node_modules/graphql/execution/execute.js:104:14)",
            "    at Object.execute (/users/user/path/to/project/node_modules/graphql/execution/execute.js:64:35)",
            "    at /users/user/path/to/project/node_modules/apollo-server-core/src/requestPipeline.ts:479:22",
            "    at Generator.next (<anonymous>)",
            "    at /users/user/path/to/project/node_modules/apollo-server-core/dist/requestPipeline.js:8:71",
            "    at new WrappedPromise (/users/user/path/to/project/node_modules/async-listener/es6-wrapped-promise.js:13:18)",
            "    at __awaiter (/users/user/path/to/project/node_modules/apollo-server-core/dist/requestPipeline.js:4:12)",
            "    at execute (/users/user/path/to/project/node_modules/apollo-server-core/dist/requestPipeline.js:220:20)",
            "    at Object.<anonymous> (/users/user/path/to/project/node_modules/apollo-server-core/src/requestPipeline.ts:366:30)",
            "    at Generator.next (<anonymous>)",
            "    at fulfilled (/users/user/path/to/project/node_modules/apollo-server-core/dist/requestPipeline.js:5:58)",
            "    at propagateAslWrapper (/users/user/path/to/project/node_modules/async-listener/index.js:504:23)"
          ]
        }
      }
    }
  ],
  "data": {
    "currentMember": null
  }
}
```

The integration looks like the following:

```js
const apolloMiddleware = apolloServer.getMiddleware({
  path: "/graphql",
});

const app = express();

// Honey badger request handler must go first
app.use(honeybadger.requestHandler);
app.use(scout.expressMiddleware());

app.use(apolloMiddleware);
app.get("/", (req, res) => res.send("Go check /graphql"));

// Honey badger request handler must go last
app.use(honeybadger.errorHandler);

app.listen({ port: PORT, host: HOST }, () => {
  console.log(`🚀 Server ready at http://${HOST}:${PORT}${apolloServer.graphqlPath}`);
});
```

Reported by [@Sandy in the scout-help channel](https://scout-community.slack.com/archives/C5X9JK1U5/p1586815796032700)

**Comment by vados-scout** [comment]
Some more details:

> 1. When a query alone, outside of the request context, there didn't seem to be an issue.
>  2. The first request seemed to hang and the second had this error.

**Comment by vados-scout** [comment]
After working on this it looks like I may have found a solution but I want to note the issues I ran into somewhere:

1) `sequelize` uses multiple `query` calls for a single raw query -- it does stuff like setting timezone, etc. 

2) Multiple `query` calls being run in quick succession revealed a bug in the `instrument()` call's handling of async namespaces -- specifically in the case where the parent of an instrumentation is *a different span*. 

When the parent of an instrumentation is a different span (i.e. a different instrumentation, not a request) we propagate the parent -- basically telling the next new span to use the same parent the span that just finished used -- but this *wasn't* working because the `doneFn` was not bound to the same namespace as the callback.

3) Since the async namespace key wasn't cleared properly, the next instrument call saw the *old* span (ex. the first `query()` performed by sequelize) as the "current" span

4) `instrument` attempts to create a child span on the "current" span (treating it as a parent) -- this immediately fails because `startChildSpan` does not allow on a closed span (this happened in (2)).

5) The `pg` integration was failing because `instrument` failed, and the object that was returned was not valid.

The fix and make this easier to catch in the future:

- Fix the namespace usage in `instrument()`
- Add a warning & some errors for this case

**Comment by vados-scout** [comment]
This exposed a subtle bug, which I will detail here for posterity:

- `cls-hooked` is less buggy and has better backwards and forwards support than `continuation-local-storage`
- first issue was that child spans were failed to be created because of the wrong context @ doneFn execution time on the span
- the first issue seemed solved (previous comment) but revealed a second deeper issue: the stopping logic that clears the async namespace needed to be attached to the request and/or spans themselves because we do not know from which async context the request/span will be stopped.

Example case: When a span that is the only one (via a single `instrument()` call) is closed, the request that was auto-created to encapsulate it needs to be closed. Up until now that was done when the `doneFn` got called (at the end of an encapsulating `scout.transaction()`/`scout.instrument()`), but it would call request.finishAndSend() -- which did *not* close the async context, it merely ended the request -- requests do not track the async context at all (and arguably shouldn't), `scout.transaction()`/`scout.instrument()` do.

Adding an `onStop` handler for requests and spans and making sure the async context is closed there shifts paradigm to ensure that async context is wiped whenever the request itself is stopped

Somewhat separately the code for handling "automatically created requests" was wrong, because it used `startTransactionSync` it needs to keep track of async contexts, and the code should have used `scout.transaction()`, *not* manually starting a transaction.

There also looks to be a bug in the tests -- when the express tests are run together, they pass but hang, tracking down what is hanging up `express` and/or the `pug` integration (it's been narrowed down to when the `pug` test and the `ejs` (or `mustache`) test run together.

**Comment by vados-scout** [comment]
Fixed (with added test) in v0.1.9 which was just released
